AOC-Day04

lahernandezb opened the pull request.

Commit:
docs: inital commit, add problem.md

Add problem prompt for day04

Commit:
feat: Add solution to part 1

Generate an array and fill connsecutively with numbers spanning input range with generateArray.js. Filter the array based on the three prompt contraints:
  1. Password should be 6 digits in length
  2. Must contain one adjacent duplicate
  3. Digits should be in increaasing order left to right

Commit:
feat: add for loop solution to part 1

Add a four loop solution in order to compare perforamnce between the for loop and array filtering solution.

Commit:
style: add liniting configuration

Commit:
test:  add mutation  testing with stryker

and relocate test files into test/ in the root directory.

Commit:
test: add testing new testing for day 04 part 1

Add testing for utlitly generateArray.js and day04.test.js

Commit:
docs: update exercise prompt to include pt.2

silverli commented on day04/generateArray.js:
```diff
@@ -0,0 +1,2 @@
+exports.generateArray = (start, end) =>
+  Array.from({ length: end + 1 - start }, (x, i) => i + start);
```
I think this `end +1 - start` is just equal to `end`!

silverli replied:
is there a ramda function that will generate a range like this?

lahernandezb replied:
There is, thanks for the tip. R.range does exactly this.

Commit:
refactor: abstract passwordFinder functionality

All index#.js files were renamed to passwordFinder#.js. New index.js will be the main entry point for the module.The passwordFinder files encapsulate functionality for the various implementations of the day 04 part 1 solution.

Abstract the regex in conditionCheck from the main module, and create isConsecutiveDuplicate utility.

Rename the num parameter of conditionCheck to numString throught  the function signature as this a more accurate description of the argument type.

Commit:
test: update stryker config

Update stryker.conf.js to ignore the interfacing js files.

Commit:
fix: passwordFinder2 returns array of numStrings

Commit:
test: Update testing for passwordFinders

Increase range and test for values of resulting array instead of testing for an empty array.

Commit:
refactor: use R.range for generateArray

Ramda's range method is more readable than the Array.from callback fn.

silverli commented on utils/isConsecutiveDuplicate.js:
```diff
@@ -0,0 +1,2 @@
+// matches all consecutive duplicates in string argument.
+exports.isConsecutiveDuplicate = numString => /(.)\1/.test(numString);
```
yay!! I love this

silverli commented on day04/passwordFinder1.js:
```diff
@@ -0,0 +1,15 @@
+  isConsecutiveDuplicate(numString) &&
+  numString === [...numString].sort(sortByNumericValue).join('');
+
+exports.passwordFinder1 = (start, end) =>
```
nice, this looks great

silverli commented on day04/index.js:
```diff
@@ -0,0 +1,10 @@
+// Retrieve acceptable password options by calling either methods with a start and end value.
+
+// getPasswordsFilter filters through a generated array of all possibilites within range.
+const getPasswordsFilter = passwordFinder1;
```
so instead of doing this, you can name the import at the top whatever you want, so you could name it getPasswordsFilter up there!

lahernandezb replied:
👍  thanks!!

silverli commented on day04/generateArray.js:
```diff
@@ -0,0 +1,3 @@
+const { range } = require('ramda');
+
+exports.generateArray = (start, end) => range(start, end + 1);
```
so much cleaner looking 👏

silverli commented on day04/index.js:
```diff
@@ -0,0 +1,10 @@
+// getPasswordsFilter filters through a generated array of all possibilites within range.
+const getPasswordsFilter = passwordFinder1;
+
+// getPasswordsLoop leverages a for loop to generarte the array of acceptable passwords.
```
mention performance stuff if you want!

silverli replied:
although one side note: I generally prefer the comments for a function like with that function in the file itself

silverli replied:
*I can't seem to edit my comment, but I meant: "although one side note: I generally prefer the comments explaining a function to go in the same file alongside the actual function logic"

silverli commented on test/day04.test.js:
```diff
@@ -0,0 +1,52 @@
+const { expect } = require('chai');
+const { passwordFinder1 } = require('../day04/passwordFinder1');
+const { passwordFinder2 } = require('../day04/passwordFinder2');
```
This is so awesome you wrote tests!!!! Maybe when you do your preparatory refactoring, you can do some Test driven development as you go! Although since most of your code is written maybe that's better suited to an exercise next time!

Commit:
refactor: rename passwordFinder imports

Import passwordFinders as getPasswordFilter and getPasswordLoop to eliminate their assignment into another variable.

Commit:
chore: add editor config to gitignore

Commit:
feat: add solution to day04 part 2

Implementation involves adding a check to our condition check function. isGroupOfTwo creates cache for each numString and stores the number of times each individual digit is present within the numString. The cache is then traversed searching for the presence of any given digit which only appears two times.

Commit:
feat: add imperative loop solution to day04 part 2

Implemenation is similar to part2-filter.js with the exception of utilizing a for loop instead of a data pipe.

Commit:
refactor: abstract reused code in parts 1 and 2

Create day04/utils directory and ...
  - Abstract the following into their own module:
    -toStringsArray
    -sortByNumericValue
    -conditionChecker
  - Create isOfLengthAndOrder module. It contains the reused conditions for part 1 and part 2 (is num length 6 and are they in increasing order).
  - Relocate generateArray into day04/utils/
  - conditionChecker is modified to a curried function. It has as parameters the 3rd conditional of this exercise which will vary depending on whether a part 1 or part 2 password is being generated (isConsecutiveDuplicates vs isGroupOfTwo).

Rename passwordFinder# files to duplicatesPasswordFinder#, and part2 files to doublePasswordFinder#

Commit:
fix: remove isADouble and use inline Fn for filter

-Update tests to include part 2 testing.
-Remove console.logs from files
-Convert part 2 exports to defualt exports.

Commit:
style: remove comments on day04/index.js

Commit:
test: add hmtl reports to stryker config

Commit:
refactor: colocate isOfLengthAndOrder and sortBy..

sortByNumericValue does not need to be its own module since it is always used with is isOfLengthAndOrder

Commit:
fix: create test suite for each file from day04

issue: stryker was reporting false mutant surviors.

solution: separating each file in day04 into its own test suite appears to solve the issue.
